### DISCUSSION TITLE
Use build-in winapi CNG

### DISCUSSION THREAD

**Discussion by 0x4v0c4d0** [kareldonk/QuantumGate#20] [Ideas]
What do you think about use build-in winapi [bcrypt](https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/) and/or ncrypt ibrary as a part of [CNG](https://docs.microsoft.com/en-us/windows/win32/seccng/about-cng) for crypt solutions instead of 3 different libs (openssl, pcg_random, monocypher) or part of them.

Bcrypt lib is rich enough. There is a [cryptograpy-grade randomizer](https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptgenrandom). And there are a lot of AES: AES-GCM, AES-GMAC, AES-CBC. ECDH, and much more. Plus ncrypt. So CNG can be replace openssl, pcg_random and monocypher.

This will make the code easier to maintain. Remove dependencies. Makes it easier to build a project. And make executable files lightweight.

Probably [the list of bcrypt supported algorithms](https://docs.microsoft.com/en-us/windows/win32/seccng/cng-algorithm-identifiers)

[There samples](https://github.com/microsoft/Windows-classic-samples/blob/main/Samples/Security/SecretAgreementWithPersistedKeys/cpp/SecretAgreementWithPersistedKeys.cpp). And aditional can be googled.

**Comment by kareldonk** [comment]
The reason why I used OpenSSL instead is because I want to support Linux in the future and would need the same API supported there as well.
